go/runtime/host/protocol/types.go

```diff
@@ -409,3 +462,20 @@ type HostFetchConsensusBlockRequest struct {
 type HostFetchConsensusBlockResponse struct {
 	Block consensus.LightBlock `json:"block"`
 }
+
+// HostFetchTxBatchRequest is a request to host to fetch a further batch of transactions. The host
+// will reply with the transactions that it hasn't sent yet.
+type HostFetchTxBatchRequest struct {
+	// Offset specifies the offset of the batch within the transaction pool.
+	//
+	// Note that this may miss transactions that have arrived after the last batch was fetched.
```

Comment:
Hm, offset doesn't seem all that useful as this might actually cause new low priority transactions to be prioritized (as high priority ones will be put in front of the queue and being missed by the offset). 

I guess it's up to the runtime to be aware of that and not use offset if it is not desired. But is there a specific case where offset would be useful?

Reply:
Yes, the idea is to support streaming transactions from the queue so that we don't need to send the whole queue into the runtime only to have a few transactions actually scheduled. So the runtime sets an initial batch size to fetch and then can use this method to request additional transactions. One could add some sort of snapshotting feature to ensure consistent reads, but not sure if this will cause any problems in practice?

Reply:
Would in that case make more sense to use a fifo queue on the scheduler? And have runtime order these by priority? That way, using offset won't miss new transactions.

Reply:
The problem in that case is that this would require sending the full queue into the runtime which this is trying to avoid. 

Missed transactions are not really a problem as they will be included in the next block (as they would be if we sent the complete queue immediately). One downside of having an integer offset is that this could result in the host sending transactions that the runtime has already seen (because new transactions were inserted above, pushing others down). 

Maybe it would be better if the offset was a transaction hash.

Reply:
Back to:

> Yes, the idea is to support streaming transactions from the queue so that we don't need to send the whole queue into the runtime only to have a few transactions actually scheduled. So the runtime sets an initial batch size to fetch and then can use this method to request additional transactions

So either:
- initial runtime request gets low number of txes -> there's not a lot of benefit from having `offset`
- initial runtime requests gets enough transactions to fill up the batch -> no need to subsequent requests (`offset` not needed)
- initial runtime requests gets a lot of transactions (but not enough to fill up the whole batch), so it tries to fetch new transactions by using `offset=<nr.of transactions>`:
  - if `limit` for the subsequent fetch is low (e.g. number of missing transactions to fill up the batch): most likely new some new high. prio transactions pushed the existing transactions, so the runtime will get a lot of duplicates and needing to do more subsequent fetches (using `transaction hash` for offset would fix this)
  - if `limit` for the subsequent is big (e.g. batch size): no benefit of using `offset` again, and it would be better to just re-fetch with `offset=0`

Reply:
Changed offset to be a hash.

Reply:
Great, thanks!